doc/plutus-core-spec/builtins-future.tex

```diff
@@ -0,0 +1,413 @@
+
+\paragraph{Uncompression for $G_1$ elements.}  Given a bytestring $b$, it is checked that
+$b$ contains exactly 48 bytes.  If not, then $\uncompress_{G_1}(b) = \errorX$ (ie,
+uncompression fails).  If the length is equal to 48 bytes, write $b$ as a
+sequence of bits: $b = b_{383} \cdots b_0$.
+\begin{itemize}
+\item If $b_{383} \neq 1$, then $\uncompress_{G_1}(b) = \errorX$.
+\item If $b_{383} = b_{382} = 1$ then
+$\uncompress_{G_1}(b) =
+\begin{cases}
+\mathcal{O}_{G_1} & \text{if $b_{381} = b_{380} = \cdots = b_0 = 0$}\\
+\errorX & \text{otherwise}.
+\end{cases}$
+\item If $b_{383}=1$ and $b_{382}=0$, let $c=\sum_{i=0}^{380}b_i2^i \in \N$.
+\begin{itemize}
+\item If $c \geq q$, $\uncompress_{G_1}(b) = \errorX$.
+\item Otherwise, let $x = \bar{c} \in \Fq$ and let $z = x^3+4$. If $z$ is not a
+square in $\Fq$, then $\uncompress_{G_1}(b) = \errorX$.
+\item If $z$ is a square then let
+$y=\begin{cases}
+\ymin(x) & \text{if $b_{381}=0$}\\
+\ymax(x) & \text{if $b_{381}=1$}.
+\end{cases}$
+\item Then $\uncompress_{G_1}(b) = \begin{cases}
+(x,y) & \text{if $(x,y) \in G_1$}\\
+\errorX & \text{otherwise}.
+\end{cases}$
+\end{itemize}
+\end{itemize}
+
+
+\paragraph{Uncompression for $G_2$ elements.}  Given a bytestring $b$, it is checked that
+$b$ contains exactly 96 bytes.  If not, then $\uncompress_{G_2}(b) = \errorX$ (ie,
+uncompression fails).  If the length is equal to 96 bytes, write $b$ as a
+sequence of bits: $b = b_{767} \cdots b_0$.
+\begin{itemize}
+\item If $b_{767} \neq 1$, then $\uncompress_{G_2}(b) = \errorX$.
+\item If $b_{767} = b_{766} = 1$ then $\uncompress_{G_2}(b) =
+\begin{cases}
+\mathcal{O}_{G_1} & \text{if $b_{765} = b_{764} = \cdots = b_0 = 0$}\\
+\errorX & \text{otherwise}.
+\end{cases}$
+\item If $b_{767}=1$ and $b_{766} = 0$, let $c=\sum_{i=0}^{383}b_i2^i$ and $d=\sum_{i=384}^{764}b_i2^{i-384} \in \N$.
+\begin{itemize}
+\item If $c \geq q$ or $d \geq q$, $\uncompress_{G_2}(b) = \errorX$.
+\item Otherwise, let $x = \bar{c}+\bar{d}u \in \Fqq$ and let $z = x^3+4(u+1)$.
+If $z$ is not a square in $\Fqq$, then $\uncompress_{G_2}(b) = \errorX$.
+\item If $z$ is a square then let
+$y=\begin{cases}
+\ymin(x) & \text{if $b_{765}=0$}\\
+\ymax(x) & \text{if $b_{765}=1$}.
+\end{cases}$
+\item Then $\uncompress_{G_2}(b) = \begin{cases}
+(x,y) & \text{if $(x,y) \in G_2$}\\
+\errorX & \text{otherwise}.
+\end{cases}$
+\end{itemize}
+\end{itemize}
+
+
+\note{Concrete syntax for BLS12-381 types.}
+\label{note:bls-syntax}
+Concrete syntax for the $\ty{bls12\_381\_G1\_element}$ and
+$\ty{bls12\_381\_G2\_element}$ types is provided via the compression and
+decompression functions defined in Notes~\ref{note:group-compression}
+and~\ref{note:group-uncompression}.  Specifically, a value of type
+$\ty{bls12\_381\_G1\_element}$ is denoted by a term of the form \texttt{(con
+bls12\_381\_G1\_element 0x...)} where \texttt{...}  consists of 96 hexadecimal
+digits representing the 48-byte compressed form of the relevant point.
```

Comment:
In the concrete syntax, does case matter for the hexadecimal digits?

Reply:
> In the concrete syntax, does case matter for the hexadecimal digits? 

No.  That should be covered by Table 6 (although I see now that I didn't say what `{48}` and `{96}` are supposed to mean: maybe we need an explanation of our syntax for syntax.).